File: R/successions.R
#' @name Successions
#' @rdname successions
#' 
#' @title Count successions
#' 
#' @description Counts the lengths of successions of identical values in a vector.
#' 
#' @param x An atomic vector.
#'
#' @return A list containing the indices of the successions (\code{index}), the total number of successions (\code{successions}), the unique value of each succession (\code{value}) and the lengths of the successions (\code{length})
#'
#' @examples
#' set.seed(7)
#' x <- sample(LETTERS[1:3], 10, replace=TRUE, prob = c(0.2, 0.2, 0.6))
#' x
#' successions(x)
#'
#' @useDynLib temporal
#' @importFrom methods getPackageName
#' @export
successions <- function (x) {
  n <- length(x)
  xdouble <- if (typeof(x) == "double") 1 else 0
  xint <- if (typeof(x) == "integer") 1 else 0
  if (!inherits(x, "factor")) x <- factor(x, levels = unique(x))
  out <- .Fortran("successions",
                  x = as.integer(x),
                  n = n,
                  z = rep(as.integer(NA), n),
                  m = rep(as.integer(NA), n),
                  PACKAGE = getPackageName(),
                  NAOK = TRUE)
  s <- sum(!is.na(out$z))
  out <- list(index = 1:s,
              successions = s,
              value = levels(x)[out$z[1:s]],
              length = as.vector(out$m[1:s]))
  if (xdouble) out$value <- as.numeric(out$value)
  if (xint) out$value <- as.integer(out$value)
  return(out)
}

File: src/successions.f90
SUBROUTINE successions(X, N, G, C)

  ! X: Group observations, e.g. group name, number etc. (it is assumed that the
  !    observations are correctly sorted)
  ! N: The number of observations, i.e. the cardinality of x.
  ! G: The unique X value per succession
  ! C: The counts of each succession
  IMPLICIT NONE
  INTEGER:: N
  INTEGER, DIMENSION(N):: X, G, C

  ! i: Looping index
  ! s: Succession counter
  ! current: Placeholder for the current value when looping through X
  ! previous: Placeholder for the previous value when looping through X
  INTEGER:: i, s, current, previous

!  ! Fill index vektor.
!  DO i = 1, N
!    A(i) = i
!  END DO

  s = 1
  G(s) = X(1)
  C(1) = 1
  DO i = 2, N
    current = X(i)
    previous = X(i - 1)

    IF (current == previous) THEN
      ! Increment counter as current observation is identical to previous
      C(s) = C(s) + 1
    ELSE
      s = s + 1
      G(s) = X(i)
      C(s) = 1
    END IF
    
  END DO


END SUBROUTINE successions


!based on either index or diff(index),
!algorithm that groups per x into sucessive groups,
!returns number of groups per x,
!amount in each group



!set.seed(29)
!df <- data.frame(i = 1:100,
!                 x = sample(LETTERS[1:3], 100, replace = TRUE),
!                 y = rpois(100, 5))


